sui_types/src/serialize.rs

```diff
@@ -12,6 +13,25 @@ use serde::{Deserialize, Serialize};
 #[path = "unit_tests/serialize_tests.rs"]
 mod serialize_tests;
 
+// SerializedEnvelope wraps application level payloads
+// (SerializedMessage) with metadata. We call the metadata
+// "headers" since they serve the same role as http headers
+// in gRPC or other http-based systems.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct SerializedEnvelope {
+    pub headers: Option<Box<SerializedHeaders>>,
+    pub payload: Box<SerializedMessage>,
```

Comment:
I don't think you need to box either `SerializedHeaders` nor `SerializedMessage`.  `SerializedMessage` already has most of its contents boxed, so all this does is force another tiny allocation of an enum int and a pointer to the boxed serializedMessage insides.   In other words we pay an allocation overhead for nothing.
Also serializedHeaders contains a Vec, which again has allocated array.

Reply:
I'm just following the existing pattern in the code. The reason it is done this way by SerializedMessage is that otherwise, it is very difficult to do the serialization without a lot of clone()ing.